---
title: "PRS Agreement Tables"
author:  "Jennifer Collister"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  word_document:
      reference_docx: "K:/TEU/TEU_Guides/TEU_DocStyle_Rmd_2020.dotx"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(kableExtra)
library(markdown)
library(knitr)
library(yaml)
library(here)
library(stringr)
library(dplyr)
library(gtools)
library(ggplot2)
library(pROC)
library(survival)
library(nricens)

# Specify the markdown format for knitr tables, otherwise they're not compatible with kableExtra
options(knitr.kable.NA='', knitr.table.format = "markdown")
options(kableExtra.auto_format = FALSE)
# Set the root directory to the project directory (otherwise knitr works in the directory of the Rmd file by default)
knitr::opts_knit$set(root.dir = here::here())

# Load the project config file for filepaths etc
config = yaml.load_file(here::here("./config.yml"))

# Source function
source(here::here(config$functions$JC))

knitr::opts_chunk$set(fig.width=9, fig.height=4)

# Specify if these tables are for the main paper or the supplementary materials
# Main paper uses 1% categories, supplementary uses 5%
supplementary <- FALSE

pct_cat <- function(var, name, pct=ifelse(supplementary==FALSE, 1, 5)) {
  cut(var, 
      breaks=c(0, pct, 20, 40, 60, 80, 100-pct, 100),
      labels=c(paste0("< ", pct, "%"), paste0(pct, "-20%"), 
               "20-40%", paste0(name, ": 40-60%"), "60-80%", 
               paste0("80-", 100-pct, "%"), paste0(">= ", 100-pct, "%"))
      )
  }

table_number <- 1

```

```{r dataset-derivation, include=FALSE, cache=FALSE, eval=FALSE}
source(file.path(config$scripts$cleaning, "dataset_generator.R"))

# Exclusion Criteria
exclusions <- function(data) {
  
  excl$initial <<- nrow(data)
  
  # Exclude those outside the 40-70 age range
  data <- data[data$TEU_BaC_AgeAtRec >= 40 & data$TEU_BaC_AgeAtRec < 70,]
  
  excl$agerange <<- nrow(data)
  
  # Restrict population to those for whom we have PRS scores available
  data <- data[!is.na(data$TEU_BrCa_313_PRS),]
  data <- data[!is.na(data$TEU_SBP_PRS),]
  
  excl$genetics <<- nrow(data)
  
  # Exclude genetic non-white groups
  data <- data[!is.na(data$GeP_ethnic),]
  excl$whitebrit <<- nrow(data)
  
  
  return(data)
}

excl <- list(initial=0)

html_header <- paste0( 
    paste0("<head><title>", "(Clifton, Lei et al. 2022)", "</title></head>"),
    paste0("<h1>", "PRS Agreement Data Dictionary", "</h1>"),
    paste0("<p>This data dictionary accompanies our paper: <i> Assessing agreement between different polygenic risk scores in the UK Biobank ",  
           text_spec("(Clifton, Lei et al. 2022)", link = "https://www.medrxiv.org/content/10.1101/2022.02.09.22270719v1"), 
           "</i></p>"), 
    paste0("<p>Please also see our ", text_spec("GitHub repository", link = "https://github.com/2cjenn/AgrPRS"), ".</p>")
    )

# Load data
data <- derive_variables(
  database = config$data$database,
  field_definitions = TEU_SPECS$PRS_agreement,
  exclusions = exclusions,
  dictionary = file.path(config$data$dictionary, "index.html"),
  html_header = html_header,
  print_derivation = TRUE,
  hide_n = TRUE
)

backup <- data

saveRDS(data, "Data/Derived/AgrPRS_20220608.rds")

```

```{r}
data <- readRDS("Data/Derived/AgrPRS_20220608.rds")
```

# Populations and PRS

All outcome specific populations are restricted to those who have genetic data available, passed genetic QC and are of White British ancestry.

## Breast cancer

```{r}

brca <- data %>% 
  filter(
    BaC_Sex=="Female",
    is.na(BrCaDx_Prevalent),
    is.na(BrCaInSitu_Prevalent),
    TEU_HES_Mast_prev=="No"
    ) %>% mutate(
      pct_mavaddat = as.numeric(quantcut(TEU_BrCa_313_PRS, q=100)),
      quint_mavaddat = as.numeric(quantcut(TEU_BrCa_313_PRS, q=5)),
      pct_fritsche = as.numeric(quantcut(TEU_BrCa_100k_PRS, q=100)),
      quint_fritsche = as.numeric(quantcut(TEU_BrCa_100k_PRS, q=5)),
      cat_mavaddat = pct_cat(pct_mavaddat, name="Mavaddat"),
      cat_fritsche = pct_cat(pct_fritsche, name="Fritsche")
      )

fritsche <- read.csv(file.path(dirname(config$data$prs$brca100k), "inputs/Betas.csv"))
mavaddat <- read.csv(file.path(dirname(config$data$prs$brca313), "inputs/Betas.csv"))

overlap_brca <- dplyr::inner_join(fritsche, mavaddat, 
                             by=c("chr_pos", "chr_name", "chr_position", "effect_allele", "reference_allele"), 
                             suffix=c("_frt", "_mav"))

no_prs <- glm(TEU_Alz_Status ~ TEU_BaC_AgeAtRec,
                 family="binomial", data=brca)

crude_fritsche <- glm(TEU_BrCa_status ~ TEU_BrCa_100k_PRS + 
                      GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                    family="binomial", data=brca)
crude_mavaddat <- glm(TEU_BrCa_status ~ TEU_BrCa_313_PRS + 
                      GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                    family="binomial", data=brca)

log_fritsche <- glm(TEU_BrCa_status ~ TEU_BaC_AgeAtRec + TEU_BrCa_100k_PRS + 
                      GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                    family="binomial", data=brca)
log_mavaddat <- glm(TEU_BrCa_status ~ TEU_BaC_AgeAtRec + TEU_BrCa_313_PRS + 
                      GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                    family="binomial", data=brca)

log_fritsche_cat <- glm(TEU_BrCa_status ~ TEU_BaC_AgeAtRec + cat_fritsche +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                        family="binomial",
                        data=brca%>%mutate(cat_fritsche = relevel(cat_fritsche, "Fritsche: 40-60%")))
log_mavaddat_cat <- glm(TEU_BrCa_status ~ TEU_BaC_AgeAtRec + cat_mavaddat +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5, 
                        family="binomial",
                        data=brca%>%mutate(cat_mavaddat = relevel(cat_mavaddat, "Mavaddat: 40-60%")))

pred_fritsche <- pct_cat(as.numeric(quantcut(predict(log_fritsche, type="response"),q=100)),name="Fritsche")
pred_mavaddat <- pct_cat(as.numeric(quantcut(predict(log_mavaddat, type="response"),q=100)),name="Mavaddat")


brca_list <- list(
  disease="Breast cancer",
  data=brca,
  outcome=brca$TEU_BrCa_status,
  LD_b_in_a = 225,
  prsA=list(
    paper="[Mavaddat2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323553/)",
    betas=mavaddat,
    no_prs=no_prs,
    crude_model_cont=crude_mavaddat,
    log_model_cont=log_mavaddat,
    log_model_cat=log_mavaddat_cat,
    log_row=8,
    prs_raw=brca$TEU_BrCa_313_PRS,
    prs_cat=brca$cat_mavaddat,
    pred=pred_mavaddat
  ),
  prsB=list(
    paper="[Fritsche2020](https://www.sciencedirect.com/science/article/pii/S0002929720303207?via%3Dihub)",
    betas=fritsche,
    no_prs=no_prs,
    crude_model_cont=crude_fritsche,
    log_model_cont=log_fritsche,
    log_model_cat=log_fritsche_cat,
    log_row=8,
    prs_raw=brca$TEU_BrCa_100k_PRS,
    prs_cat=brca$cat_fritsche,
    pred=pred_fritsche
  )
)

```

Exclusions:

* Male
* Prevalent breast cancer
* Prevalent carcinoma in situ
* Mastectomy

Outcome: 

* Breast cancer diagnosis in HES or Death Registry

Breast cancer population: **n = `r nrow(brca)`**

PRS scores:

* [Mavaddat2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323553/) (**313 SNPs**)
  * PMID: [30554720](https://pubmed.ncbi.nlm.nih.gov/30554720/) 
  * PMCID: [PMC6323553](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6284793/)
  * Developed using hard-thresholding with threshold $p < 10^{-5}$
  * "*Finally, the effect sizes for the selected SNPs were jointly estimated in a single logistic regression model.*"
  * 306 of the 313 SNPs are available in UKB (This is mentioned in the Mavaddat paper as well, and in their supplementary materials, pp21-22)
  * Using data from [Michailidou2017 BCAC](https://www.nature.com/articles/nature24284)
* [Fritsche2020](https://www.sciencedirect.com/science/article/pii/S0002929720303207?via%3Dihub) (**118,388 SNPs**)
  * PMID: [32991828](https://pubmed.ncbi.nlm.nih.gov/32991828/) 
  * PMCID: [PMC7675001](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7675001/)
  * Developed using Lassosum with shrinkage parameter $s = 0.5$ and tuning parameter $\lambda = 0.0043$
  * "*Lassosum obtains PRS weights by applying elastic net penalization to GWAS summary statistics and incorporating LD information from a reference panel.*"
  * Using data from [Michailidou2017 BCAC](https://www.nature.com/articles/nature24284) and developed in Michigan Genome Initiative data (MGI)
  
## Hypertension

```{r}

htn <- data %>%
  filter(
    !is.na(TEU_BlP_SBP.avg),
    TEU_BlP_nSBP == 2,
    TEU_BlP_SBP.avg >= 70 & TEU_BlP_SBP.avg <= 270,
    TEU_MACE_prev=="No",
  )%>%
  mutate(
    pct_evangelou = as.numeric(quantcut(TEU_SBP_PRS, q=100)),
    quint_evangelou = as.numeric(quantcut(TEU_SBP_PRS, q=5)),
    pct_warren = as.numeric(quantcut(TEU_SBP_PRS_Warren, q=100)),
    quint_warren = as.numeric(quantcut(TEU_SBP_PRS_Warren, q=5)),
    cat_warren = pct_cat(pct_warren, name="Warren"),
    cat_evangelou = pct_cat(pct_evangelou, name="Evangelou"),
    TEU_HTN_status = as.numeric(TEU_BlP_SBP.avg >= 140)
    )


evangelou <- read.csv(file.path(dirname(config$data$prs$sbp_evangelou), "inputs/Betas.csv")) %>% filter(!is.na(Beta_SBP))
warren <- read.csv(file.path(dirname(config$data$prs$sbp_warren), "inputs/Betas.csv"))

overlap_htn <- dplyr::inner_join(evangelou, warren, 
                             by=c("chr_pos", "chr_name", "chr_position"), 
                             suffix=c("_eva", "_war")) %>%
  mutate(Beta_SBP_war=ifelse(effect_allele_eva==reference_allele_war, -Beta_SBP_war, Beta_SBP_war))

no_prs <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec,
                 family="binomial", data=htn)

crude_evangelou <- glm(TEU_HTN_status ~ TEU_SBP_PRS +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                     family="binomial", data=htn)
crude_warren <- glm(TEU_HTN_status ~ TEU_SBP_PRS_Warren +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=htn)

log_evangelou <- glm(TEU_HTN_status ~ BaC_Sex + TEU_BaC_AgeAtRec + TEU_SBP_PRS +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                     family="binomial", data=htn)
log_warren <- glm(TEU_HTN_status ~ BaC_Sex + TEU_BaC_AgeAtRec + TEU_SBP_PRS_Warren +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=htn)

log_evangelou_cat <- glm(TEU_HTN_status ~ BaC_Sex + TEU_BaC_AgeAtRec + cat_evangelou +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                         family="binomial",
                        data=htn%>%mutate(cat_evangelou = relevel(cat_evangelou, "Evangelou: 40-60%")))
log_warren_cat <- glm(TEU_HTN_status ~ BaC_Sex + TEU_BaC_AgeAtRec + cat_warren +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                      family="binomial",
                        data=htn%>%mutate(cat_warren = relevel(cat_warren, "Warren: 40-60%")))

pred_warren <- pct_cat(as.numeric(quantcut(predict(log_warren, type="response"),q=100)),name="Warren")
pred_evangelou <- pct_cat(as.numeric(quantcut(predict(log_evangelou, type="response"),q=100)),name="Evangelou")


htn_list <- list(
  disease="Hypertension",
  data=htn,
  outcome=htn$TEU_HTN_status,
  LD_b_in_a = 210,
  prsA=list(
    paper="[Warren2017](https://www.nature.com/articles/ng.3768)",
    betas=warren,
    no_prs=no_prs,
    crude_model_cont=crude_warren,
    log_model_cont=log_warren,
    log_model_cat=log_warren_cat,
    log_row=9,
    prs_raw=htn$TEU_SBP_PRS_Warren,
    prs_cat=htn$cat_warren,
    pred=pred_warren
  ),
  prsB=list(
    paper="[Evangelou2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6284793/)",
    betas=evangelou,
    no_prs=no_prs,
    crude_model_cont=crude_evangelou,
    log_model_cont=log_evangelou,
    log_model_cat=log_evangelou_cat,
    log_row=9,
    prs_raw=htn$TEU_SBP_PRS,
    prs_cat=htn$cat_evangelou,
    pred=pred_evangelou
  )
)

```

Exclusions:

* Missing or fewer than 2 BP measurements
* Implausible BP measurements
* Prevalent MACE in HES, using ICD codes from Neo's HTN/Chol MACE paper

Outcome:

* Average measured systolic BP at baseline >= 140 mmHg

Hypertension population: **n = `r nrow(htn)`**

PRS Scores:

* [Warren2017](https://www.nature.com/articles/ng.3768) (**267 SNPs**)
  * PMID: [28135244](https://pubmed.ncbi.nlm.nih.gov/28135244/) 
  * PMCID: [PMC5972004](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5972004/)
  * "*We created weighted GRSs for all [...] previously reported variants at the time of analysis and our validated variants*"
* [Evangelou2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6284793/) (**884 SNPs**)
  * PMID: [30224653](https://pubmed.ncbi.nlm.nih.gov/30224653/) 
  * PMCID: [PMC6284793](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6284793/)
  * While UKB was included in some of the GWAS where the SNPs we identified, we've previously determined that the UKB data was not used to determine the betas
  * There are 885 SNPs in the table from Evangelou, but one of them doesn't have a beta for SBP (just for DBP and pulse pressure) so the SBP PRS only contains 884 SNPs.

  
## Dementia

```{r}

alz <- data %>%
  filter(
    TEU_HES_Alz_prev == "No" # Uses ICD codes from Tom's dementia htn paper
  ) %>%
  mutate(
    pct_ebenau = as.numeric(quantcut(Alz_PRS_Ebenau, q=100)),
    quint_ebenau = as.numeric(quantcut(Alz_PRS_Ebenau, q=5)),
    pct_najar = as.numeric(quantcut(Alz_PRS_Najar, q=100)),
    quint_najar = as.numeric(quantcut(Alz_PRS_Najar, q=5)),
    cat_ebenau = pct_cat(pct_ebenau, name="Ebenau"),
    cat_najar = pct_cat(pct_najar, name="Najar"),
  )

ebenau <- read.csv(file.path(dirname(config$data$prs$alz_ebenau), "inputs/Betas.csv"))
najar <- read.csv(file.path(dirname(config$data$prs$alz_najar), "inputs/Betas.csv"))

overlap_alz <- dplyr::inner_join(najar, ebenau, 
                             by=c("chr_name", "chr_position"), 
                             suffix=c("_naj", "_ebe"))

no_prs <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec,
                 family="binomial", data=alz)

crude_ebenau <- glm(TEU_Alz_Status ~ Alz_PRS_Ebenau +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=alz)
crude_najar <- glm(TEU_Alz_Status ~ Alz_PRS_Najar +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=alz)


log_ebenau <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec + Alz_PRS_Ebenau +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=alz)
log_najar <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec + Alz_PRS_Najar +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                  family="binomial", data=alz)



log_ebenau_cat <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec + cat_ebenau +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                      family="binomial",
                        data=alz%>%mutate(cat_ebenau = relevel(cat_ebenau, "Ebenau: 40-60%")))
log_najar_cat <- glm(TEU_Alz_Status ~ BaC_Sex + TEU_BaC_AgeAtRec + cat_najar +
                           GeP_Array + GeP_PC_1 + GeP_PC_2 + GeP_PC_3 + GeP_PC_4 + GeP_PC_5,
                      family="binomial",
                        data=alz%>%mutate(cat_najar = relevel(cat_najar, "Najar: 40-60%")))


pred_ebenau <- pct_cat(as.numeric(quantcut(predict(log_ebenau, type="response"),q=100)),name="Ebenau")
pred_najar <- pct_cat(as.numeric(quantcut(predict(log_najar, type="response"),q=100)),name="Najar")

alz_list <- list(
  disease="Dementia",
  data=alz,
  outcome=alz$TEU_Alz_Status,
  LD_b_in_a = 13,
  prsA=list(
    paper="[Najar2021](https://alz-journals.onlinelibrary.wiley.com/doi/10.1002/dad2.12142)",
    betas=najar,
    no_prs=no_prs,
    crude_model_cont=crude_najar,
    log_model_cont=log_najar,
    log_model_cat=log_najar_cat,
    log_row=9,
    prs_raw=alz$Alz_PRS_Najar,
    prs_cat=alz$cat_najar,
    pred=pred_najar
  ),
  prsB=list(
    paper="[Ebenau2021](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8438688/)",
    betas=ebenau,
    no_prs=no_prs,
    crude_model_cont=crude_ebenau,
    log_model_cont=log_ebenau,
    log_model_cat=log_ebenau_cat,
    log_row=9,
    prs_raw=alz$Alz_PRS_Ebenau,
    prs_cat=alz$cat_ebenau,
    pred=pred_ebenau)
)


```

Exclusions:

* Prevalent Alzheimer's in HES, using ICD codes from Tom's dementia/htn paper

Outcome: 

* Dementia/Alzheimer's reported in HES or Death Registry

Dementia population: **n = `r nrow(alz)`**

PRS Scores:

* [Najar2021](https://doi.org/10.1002/dad2.12142) (**57 SNPs**)
  * PMID: [33532541](https://pubmed.ncbi.nlm.nih.gov/33532541/)
  * PMCID: [PMC7821873](http://www.ncbi.nlm.nih.gov/pmc/articles/pmc7821873/)
  * "*AD-PRSs were generated using summary statistics from stage 1 of the most recent AD GWAS including clinically defined AD. SNPs were selected using linkage disequilibrium (LD) clumping. All variants in the APOE region (chromosome 19, coordinates hg19 [GRCh37]: 44412079 to 46412079) were removed. In the present study, we created PRSs including variants that surpassed [...] P value thresholds [...], referred to as 1eâ€“5 AD-PRS (including 57 SNPs [...])*"
* [Ebenau2021](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8438688/) (**39 SNPs**)
  * PMID: [34541285](https://pubmed.ncbi.nlm.nih.gov/34541285/) 
  * PMCID: [PMC8438688](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8438688/)
  * Unclear how SNPs were selected - it just says: "*We selected 39 genetic variants for which there was evidence of a significant association with AD from previous genome-wide association studies (GWAS) and candidate-gene studies*"
  * Effect sizes came from IGAP GWAS [Kunkle2019](https://www.nature.com/articles/s41588-019-0358-2)


# Table 1: Summary

```{r, warning=FALSE, message=FALSE, fig.show='hide'}

row_creator <- function(prsobj, outcome) {
  row <- list(PRS=prsobj$paper,
               nSNPs=nrow(prsobj$betas),
               OR=OR_CI(prsobj$log_model_cat, row=prsobj$log_row, dp=2),
               AUC_crude=adjAUC(status=outcome, prediction=predict(prsobj$crude_model_cont, type="response"), dp=2),
               AUC_adj=adjAUC(status=outcome, prediction=predict(prsobj$log_model_cont, type="response"), dp=2)
               )
}

row_joiner <- function(dis) {
  rowA <- append(row_creator(dis$prsA, dis$outcome),
                list(
                  disease=paste0(dis$disease, " (", format(nrow(dis$data), big.mark=",", scientific=FALSE), ")"),
                  NRI=contNRI(x1=predict(dis$prsA$log_model_cont, type="response"), 
                              x2=predict(dis$prsB$log_model_cont, type="response"),
                              y=dis$outcome, dp=2),
                  corr=pretty_dp(cor(dis$prsA$prs_raw, dis$prsB$prs_raw),2),
                  LD=paste0(dis$LD_b_in_a, " (", round(100*dis$LD_b_in_a/nrow(dis$prsA$betas)), "%)")
                  )
                )
  rowB <- append(row_creator(dis$prsB, dis$outcome),
                 list(
                   disease=NA_character_,
                   NRI=NA_character_,
                   corr=NA_character_,
                   LD=NA_integer_
                   )
                 )
  df <- do.call(rbind.data.frame, list(rowA, rowB)) %>% select(disease, PRS, nSNPs, OR, AUC_crude, AUC_adj, NRI, corr, LD)
}

tab1 <- rbind(row_joiner(brca_list),
              row_joiner(htn_list),
              row_joiner(alz_list)
              )

rownames(tab1) <- NULL
colnames(tab1) <- c("Disease (N)", "PRS", "nSNPs","OR", "Crude AUC", "Multi AUC", "NRI", "r", "LD > 0.8")

kable(tab1,
      caption=paste0("Table ", table_number, ": PRS compared for each outcome and their performance characteristics in the UK Biobank. N: number of participants whose PRS score was obtained. nSNPs: number of SNPs in PRS. OR: odds ratio for top ", ifelse(supplementary==TRUE, "5", "1"), "% vs middle quintile of PRS from multivariable logistic regression model adjusted for age, sex, genotyping array and first 5 PCs. Crude AUC: area under receiver-operating curve using predicted risk from multivariable logistic regression model containing continuous PRS, genotyping array and first 5 PCs. Multi AUC: area under receiver-operating curve using predicted risk from multivariable logistic regression model containing age, sex, continuous PRS, genotyping array and first 5 PCs. NRI: continuous net reclassification index using predicted risks from two multivariable logistic regression models containing age, sex, continuous PRS for this disease, genotyping array and first 5 PCs. The PRS with more SNPs is considered the 'updated' model. r: Pearson correlation coefficient between the two continuous PRS for this disease. LD: number (%) of SNPs in the 'updated' PRS which either appear in or are in linkage disequilbrium (R^2^ > 0.8) with SNPs in the 'original' PRS. Breast cancer models not adjusted for sex because population is restricted to females. 95% CI for AUC and NRI calculated by bootstrapping."))

table_number <- table_number + 1

```

# Cross-tabs of PRS categories

```{r}

cellpcts <- function(obj, var="prs_cat", dp=1) {
  tab <- table(obj$prsA[[var]], obj$prsB[[var]])
  
  rowpcts <- pretty_dp(100*prop.table(tab, margin=1),dp)
  colpcts <- pretty_dp(100*prop.table(tab, margin=2),dp)
  cellpcts <- pretty_dp(100*prop.table(tab),dp)
  tab <- pretty_dp(tab, dp=0)
  
  tab_str <- paste0(tab, " (", rowpcts, ", ", colpcts, ", ", cellpcts, ")")
  tab_out <- matrix(tab_str, nrow=dim(tab)[1], ncol=dim(tab)[2])
  
  colnames(tab_out) <- colnames(tab)
  rownames(tab_out) <- rownames(tab)
  
  caption <- paste0("Table ", table_number, ": Cross-classification of predicted risk of ", tolower(obj$dis), ", according to the percentiles of each PRS. Number of participants are shown as n (col%, row%, cell%). Higher percentiles of PRS indicate increased risk of ", tolower(obj$dis), ".")
  
  table_number <<- table_number + 1
  
  return(list(tab=tab_out, cap=caption))
}

```

## Breast cancer (n = `r nrow(brca)`)

```{r}

tab <- cellpcts(brca_list, var="prs_cat")
kable(tab$tab, caption=tab$cap)

```

## Hypertension (n = `r nrow(htn)`)

```{r}

tab <- cellpcts(htn_list, var="prs_cat")
kable(tab$tab, caption=tab$cap)

```

## Dementia (n = `r nrow(alz)`)

```{r}

tab <- cellpcts(alz_list, var="prs_cat")
kable(tab$tab, caption=tab$cap)

```

# Categorical NRI, cross-tab and percent reclassified among those with outcome present

```{r, messages=FALSE, warnings=FALSE, fig.show='hide'}

nri_tab <- function(obj, dp=1) {
  initial <- obj$prsA$prs_cat
  updated <- obj$prsB$prs_cat
  status <- obj$outcome
  
  tab <- table(initial[status==1], updated[status==1])
  rowpcts <- pretty_dp(100*prop.table(tab, margin=1),dp)
  
  tab_str  <- paste0(tab, " (", rowpcts, ")") 
  tab_out <- matrix(tab_str, nrow=dim(tab)[1], ncol=dim(tab)[2])
  
  colnames(tab_out) <- colnames(tab)
  rownames(tab_out) <- rownames(tab)
  
  
  tab1 <- cbind(tab_out, ` % reclassified` = paste0(pretty_dp((rowSums(tab) - diag(tab))/rowSums(tab) * 100, dp),"%"))
  
  x <- nribin_silent(event = status, 
              p.std = as.numeric(initial),
              p.new = as.numeric(updated),
              updown="category", cut=seq(1,7), niter=1000)

  
  caption <- paste0("Table ", table_number, ": Cross-classification of predicted risk of ", tolower(obj$dis), 
                    " among participants who experienced the outcome, according to the percentiles of each PRS.
                    Initial score is ", obj$prsA$paper, ", updated score is ", obj$prsB$paper, ". ", 
                     format(sum(obj$outcome), big.mark=",", scientific=FALSE), " of ", 
                    format(nrow(obj$data), big.mark=",", scientific=FALSE), " individuals (", 
                     pretty_dp(100*sum(obj$outcome)/nrow(obj$data),1), "%) experienced the outcome. ",
                    "Row %s shown, and may not add up to 100 due to rounding. ",
                    "Categorical NRI (95% CI): ", round(x$nri$Estimate[1], 3), 
                    " (", round(x$nri$Lower[1], 3), ", ", round(x$nri$Upper[1], 3), ").")
  
  table_number <<- table_number + 1
  
  return(list(tab=tab1, cap=caption))
}

```

## Breast cancer: outcome present

```{r}

nri <- nri_tab(brca_list)

kable(nri$tab, caption=nri$cap)

```

## Hypertension: outcome present

```{r}

nri <- nri_tab(htn_list)

kable(nri$tab, caption=nri$cap)

```

## Dementia: outcome present

```{r}

nri <- nri_tab(alz_list)

kable(nri$tab, caption=nri$cap)

```

# Figure 1: ROC plot

ROC plots obtained from predictions from multivariable logistic regression of age, sex and continuous PRS against disease outcome.

```{r}

rocplot <- function(obj, disease){
  
  # par(mfrow=c(1,2))

  plot.roc(obj$outcome, predict(obj$prsA$log_model_cont, type="response"),
           levels=c(0,1), direction="<",
           percent = TRUE,                    # show all values in percent
           print.auc=TRUE,
           #display pAUC value on the plot with following options:
           print.auc.pattern = "AUC %.1f%%",
           print.auc.cex=1.5,
           main = paste0(disease, ": PRS-A"))

  plot.roc(obj$outcome, predict(obj$prsB$log_model_cont, type="response"),
           levels=c(0,1), direction="<",
           percent = TRUE,                    # show all values in percent
           print.auc=TRUE,
           #display pAUC value on the plot with following options:
           print.auc.pattern = "AUC %.1f%%",
           print.auc.cex=1.5,
           main = "PRS-B")
  
  # par(mfrow=c(1,1))

}

ggrocplot <- function(obj) {
  
  plotA <- ggroc(roc(obj$outcome, predict(obj$prsA$log_model_cont, type="response"), 
          levels=c(0,1), direction="<"))
  
  plotB <- ggroc(roc(obj$outcome, predict(obj$prsB$log_model_cont, type="response"), 
          levels=c(0,1), direction="<"))
  
  return(list(plotA=plotA, plotB=plotB))
}

# test <- ggrocplot(alz_list)
# grid.arrange(test$plotA, test$plotB, ncol=2)

```


## Figures

Figure 1: ROC plots for breast cancer, hypertension and dementia, obtained from the multivariable logistic regression containing age, continuous PRS, genotyping array and first 5 PCs. 

```{r, fig.width=8.5, fig.height=10}
par(mfrow=c(3, 2))
rocplot(brca_list, disease="Breast cancer")
rocplot(htn_list, disease="Hypertension")
rocplot(alz_list, disease="Dementia")

```



```{bash, eval=FALSE}
# Count the number of SNPs in LD > 0.8 between the scores
# See K:\TEU\TEU_Members\Jennifer_Collister\bmrc-ukb\prs\projects\AgrPRS\scripts

# For each condition:

# Number of SNPs extracted, from both PRS at once
wc -l single_allelic.txt

# Number of SNPs that are in both PRS
DUP=wc -l sorted.rmdup.list

# Number of SNPs identified as "tags" (LD > 0.8)
# Note all SNPs in PRS B are identified as tags for all SNPs in PRS B
TAG=wc -l LD.tags

# Number of SNPs in PRS B
PRS_B=wc -l prsB_list.txt

# Calculate number of overlapping/LD SNPs
$DUP + $TAG - $PRS_B

# Note running these commands as written wouldn't actually work, because the output of wc -l contains the file name
# But this demonstrates the process :)

```

